### PR TITLE
fix(cleanup) : Fixed creation of cleanUp Job

### PR DESCRIPTION
**Pull request by rahulchheda**
- Added function to get Node Object using Cleaner Operator
    - Added function to get Tolerations from the Node Taints
    - Added one more parameter in NewCleanupJob to set Tolerations of Job to be created.

Signed-off-by: Rahul M Chheda <rahul.chheda@mayadata.io>

Logs: These are sequential logs, will the steps I performed:
```
rahul_chheda_mayadata_io@rahulchheda:~/openebs$ kubectl get pods -n openebs
NAME                                           READY   STATUS    RESTARTS   AGE
maya-apiserver-5865d7d548-bmzck                1/1     Running   0          2d
openebs-admission-server-684d455954-zvgrv      1/1     Running   0          2d
openebs-localpv-provisioner-58f5df5458-ttmw8   1/1     Running   0          2d
openebs-ndm-8c4cs                              1/1     Running   0          21s
openebs-ndm-dmf9f                              1/1     Running   0          21s
openebs-ndm-gw9bx                              1/1     Running   0          21s
openebs-ndm-operator-8456cf565c-kh9mk          1/1     Running   0          2m2s
openebs-provisioner-594bd6cf78-4zqf5           1/1     Running   0          2d
openebs-snapshot-operator-777cfb86df-kpzs8     2/2     Running   0          2d
rahul_chheda_mayadata_io@rahulchheda:~/openebs$ kubectl apply -f bdc.yaml -n openebs
blockdeviceclaim.openebs.io/example-blockdeviceclaim created
rahul_chheda_mayadata_io@rahulchheda:~/openebs$ kubectl get bdc -n openebs
NAME                       BLOCKDEVICENAME                                PHASE   AGE
example-blockdeviceclaim   blockdevice-860ab5d1205c0f2891d5bbfea2e223e5   Bound   10s
rahul_chheda_mayadata_io@rahulchheda:~/openebs$ kubectl get pods -n openebs
NAME                                           READY   STATUS    RESTARTS   AGE
maya-apiserver-5865d7d548-bmzck                1/1     Running   0          2d
openebs-admission-server-684d455954-zvgrv      1/1     Running   0          2d
openebs-localpv-provisioner-58f5df5458-ttmw8   1/1     Running   0          2d
openebs-ndm-8c4cs                              1/1     Running   0          51s
openebs-ndm-dmf9f                              1/1     Running   0          51s
openebs-ndm-gw9bx                              1/1     Running   0          51s
openebs-ndm-operator-8456cf565c-kh9mk          1/1     Running   0          2m32s
openebs-provisioner-594bd6cf78-4zqf5           1/1     Running   0          2d
openebs-snapshot-operator-777cfb86df-kpzs8     2/2     Running   0          2d
rahul_chheda_mayadata_io@rahulchheda:~/openebs$ kubectl taint nodes --all user=pass:NoSchedule
node/gke-standard-cluster-4-default-pool-ad827b74-jvcc tainted
node/gke-standard-cluster-4-default-pool-ad827b74-rj80 tainted
node/gke-standard-cluster-4-default-pool-ad827b74-w2j4 tainted
rahul_chheda_mayadata_io@rahulchheda:~/openebs$ kubectl delete bdc example-blockdeviceclaim -n openebs
kublockdeviceclaim.openebs.io "example-blockdeviceclaim" deleted
rahul_chheda_mayadata_io@rahulchheda:~/openebs$ kubectl get pods -n openebs
NAME                                                         READY   STATUS      RESTARTS   AGE
cleanup-blockdevice-860ab5d1205c0f2891d5bbfea2e223e5-gmn2q   0/1     Completed   0          3s
maya-apiserver-5865d7d548-bmzck                              1/1     Running     0          2d
openebs-admission-server-684d455954-zvgrv                    1/1     Running     0          2d
openebs-localpv-provisioner-58f5df5458-ttmw8                 1/1     Running     0          2d
openebs-ndm-8c4cs                                            1/1     Running     0          116s
openebs-ndm-dmf9f                                            1/1     Running     0          116s
openebs-ndm-gw9bx                                            1/1     Running     0          116s
openebs-ndm-operator-8456cf565c-kh9mk                        1/1     Running     0          3m37s
openebs-provisioner-594bd6cf78-4zqf5                         1/1     Running     0          2d
openebs-snapshot-operator-777cfb86df-kpzs8                   2/2     Running     0          2d
rahul_chheda_mayadata_io@rahulchheda:~/openebs$ kubectl get pods -n openebs
NAME                                                         READY   STATUS      RESTARTS   AGE
cleanup-blockdevice-860ab5d1205c0f2891d5bbfea2e223e5-gmn2q   0/1     Completed   0          5s
maya-apiserver-5865d7d548-bmzck                              1/1     Running     0          2d
openebs-admission-server-684d455954-zvgrv                    1/1     Running     0          2d
openebs-localpv-provisioner-58f5df5458-ttmw8                 1/1     Running     0          2d
openebs-ndm-8c4cs                                            1/1     Running     0          118s
openebs-ndm-dmf9f                                            1/1     Running     0          118s
openebs-ndm-gw9bx                                            1/1     Running     0          118s
openebs-ndm-operator-8456cf565c-kh9mk                        1/1     Running     0          3m39s
openebs-provisioner-594bd6cf78-4zqf5                         1/1     Running     0          2d
openebs-snapshot-operator-777cfb86df-kpzs8                   2/2     Running     0          2d
rahul_chheda_mayadata_io@rahulchheda:~/openebs$ 
```

### CODE REVIEW

File: pkg/cleaner/jobcontroller.go [hunk 1]
```diff
@@ -190,7 +190,7 @@ func (c *jobController) CancelJob(bdName string) error {
 	job := &batchv1.Job{}
 	err := c.client.Get(context.TODO(), objKey, job)
 
-	err = c.client.Delete(context.TODO(), job, client.PropagationPolicy(metav1.DeletePropagationForeground))
+	//err = c.client.Delete(context.TODO(), job, client.PropagationPolicy(metav1.DeletePropagationForeground))
```
Comment (prateekpandey14) [inline pkg/cleaner/jobcontroller.go]: Why this change ?

Comment (rahulchheda) [reply pkg/cleaner/jobcontroller.go]: Just for testing purpose, let me fix this.

File: pkg/cleaner/cleaner.go [hunk 2]
```diff
@@ -24,6 +24,9 @@ import (
 	"context"
 	"github.com/openebs/node-disk-manager/pkg/apis/openebs/v1alpha1"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/openebs/node-disk-manager/cmd/ndm_daemonset/controller"
```
Comment (mittachaitu) [inline pkg/cleaner/cleaner.go]: Usually, this shouldn't import i.e inside pkg structure cmd/ shouldn't be imported.

Comment (rahulchheda) [reply pkg/cleaner/cleaner.go]: Observed this from `pkg/cleanerjobcontroller.go`, let me try to get around it.

File: pkg/cleaner/cleaner.go [hunk 3]
```diff
@@ -140,13 +143,48 @@ func (c *CleanupStatusTracker) CancelJob(bdName string) error {
 
 // runJob creates a new cleanup job in the namespace
 func (c *Cleaner) runJob(bd *v1alpha1.BlockDevice, volumeMode VolumeMode) error {
-	job, err := NewCleanupJob(bd, volumeMode, c.Namespace)
+
+	//retreive node Object to pass tolerations to the Job
+	selectedNode, err := c.getNodeObjectByNodeName(bd.Labels[controller.KubernetesHostNameLabel])
```
Comment (kmova) [inline pkg/cleaner/cleaner.go]: If this is the only place that you are using `controller` to get the constant, then it is better to directly use the constant then import it. Also should this be HostName or nodeName?

Comment (kmova) [reply pkg/cleaner/cleaner.go]: cc: @akhilerm

Comment (akhilerm) [reply pkg/cleaner/cleaner.go]: This will give the hostname @kmova  and @rahulchheda . The method should be `getNodeObjectByHostName`. I think in localpv also we are using the same method. We can filter from nodes based on the hostname label.

File: pkg/cleaner/cleaner.go [hunk 4]
```diff
@@ -140,13 +143,48 @@ func (c *CleanupStatusTracker) CancelJob(bdName string) error {
...
+	tolerations := getTolerationsForTaints(selectedNode.Spec.Taints...)
+
+	job, err := NewCleanupJob(bd, volumeMode, tolerations, c.Namespace)
 	if err != nil {
 		return err
 	}
 	return c.Client.Create(context.TODO(), job)
 }
 
+func (c *Cleaner) getNodeObjectByNodeName(nodeName string) (*v1.Node, error) {
```
Comment (akhilerm) [inline pkg/cleaner/cleaner.go]: `getNodeObjectByHostName(hostName string)`
